### PR TITLE
Fix WebKit static analyzers to support ref and deref methods to be defined on different classes.

### PR DESCRIPTION
**Pull request by rniwa**

### CODE REVIEW

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 1]
```diff
@@ -77,14 +77,53 @@ class RefCntblBaseVirtualDtorChecker
               (AccSpec == AS_none && RD->isClass()))
             return false;
 
-          std::optional<const CXXRecordDecl*> RefCntblBaseRD = isRefCountable(Base);
-          if (!RefCntblBaseRD || !(*RefCntblBaseRD))
+          auto hasRefInBase = clang::hasPublicMethodInBase(Base, "ref");
```
Comment (haoNoQ) [inline clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp:80]: I wonder if this entire new section can be deduplicated with `isRefCountable()` in the other file. They look so similar!

**Comment by haoNoQ** [comment]
LGTM, thanks!! I have a couple nitpicks but I don't think they should block your progress.

**Review by haoNoQ — APPROVED** [review]

**Comment by haoNoQ** [comment]
(You still need me to merge right? You can ask me to merge, or you can address nitpicks and then ask me to merge, or you can go through https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access and merge yourself!)

File: clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp [hunk 2]
```diff
@@ -77,14 +77,53 @@ class RefCntblBaseVirtualDtorChecker
...
+          auto hasDerefInBase = clang::hasPublicMethodInBase(Base, "deref");
+
+          bool hasRef = hasRefInBase && *hasRefInBase != nullptr;
+          bool hasDeref = hasDerefInBase && *hasDerefInBase != nullptr;
+
+          const Type *T = Base->getType().getTypePtrOrNull();
+          if (!T)
+            return false;
+
+          const CXXRecordDecl *C = T->getAsCXXRecordDecl();
```
[the remark this answers is not in the document]
Comment (rniwa) [reply clang/lib/StaticAnalyzer/Checkers/WebKit/RefCntblBaseVirtualDtorChecker.cpp]: Fixed.

**Comment by rniwa** [comment]
Yes, it would be great if you can merge the change now that I've addressed your nits.

**Pull request merged**